include/pybind11/detail/common.h

```diff
@@ -735,6 +735,12 @@ using expand_side_effects = bool[];
 
 PYBIND11_NAMESPACE_END(detail)
 
+#ifdef __MINGW32__
+#  pragma push_macro("PYBIND11_EXPORT")
```

Comment:
This really worries me: it's the kind of thing that's easily overlooked, highly surprising, can be extremely confusing when this hits in the middle of complex bigger changes (refactoring), and likely to cost hours to finally realize and work around. But I want to be pragmatic:

1. Is there a way this could be avoided, for example by being smarter in the primary definition of PYBIND11_EXPORT?

2. What about something like PYBIND11_EXPORT_ALL and PYBIND11_EXPORT_SOME (with a more meaningful name)?

3. If there is really no other way, could you please add a comment to explain in detail why this is needed, what is the scope of code where this needs to be blank for MINGW, and what is the code that needs the "original" define?

Reply:
PYBIND11_EXPORT_EXCEPTION ?

Reply:
PYBIND11_EXPORT_STDLIB ?

Reply:
Yeah, that sounds like a good direction!
Looking systematically:
```
include/pybind11/pytypes.h:class PYBIND11_EXPORT error_already_set : public std::runtime_error {
include/pybind11/detail/common.h:#if !defined(PYBIND11_EXPORT)
include/pybind11/detail/common.h:#    define PYBIND11_EXPORT __declspec(dllexport)
include/pybind11/detail/common.h:#    define PYBIND11_EXPORT __attribute__ ((visibility("default")))
include/pybind11/detail/common.h:    extern "C" PYBIND11_MAYBE_UNUSED PYBIND11_EXPORT PyObject *PyInit_##name(); \
include/pybind11/detail/common.h:    extern "C" PYBIND11_EXPORT PyObject *PyInit_##name()
include/pybind11/detail/common.h:    extern "C" PYBIND11_MAYBE_UNUSED PYBIND11_EXPORT void init##name(); \
include/pybind11/detail/common.h:    extern "C" PYBIND11_EXPORT void init##name() {                      \
include/pybind11/detail/common.h:class PYBIND11_EXPORT builtin_exception : public std::runtime_error {
include/pybind11/detail/common.h:    class PYBIND11_EXPORT name : public builtin_exception { public: \
docs/advanced/misc.rst:    class PYBIND11_EXPORT Dog : public Animal {
CMakeLists.txt:  if(NOT PYBIND11_EXPORT_NAME)
CMakeLists.txt:    set(PYBIND11_EXPORT_NAME "${PROJECT_NAME}Targets")
CMakeLists.txt:  install(TARGETS pybind11_headers EXPORT "${PYBIND11_EXPORT_NAME}")
CMakeLists.txt:    EXPORT "${PYBIND11_EXPORT_NAME}"
tests/test_cmake_build/subdirectory_embed/CMakeLists.txt:set(PYBIND11_EXPORT_NAME test_export)
tests/test_cmake_build/subdirectory_embed/CMakeLists.txt:# Test custom export group -- PYBIND11_EXPORT_NAME
tests/test_exceptions.h:class PYBIND11_EXPORT shared_exception : public pybind11::builtin_exception {
tests/cross_module_gil_utils.cpp:extern "C" PYBIND11_EXPORT
```

Trying to understand:

* Are the hits in CMakeLists.txt unrelated to the C++ macro?
* `PYBIND11_EXPORT` is documented, probably it will need some care?
* There aren't that many hits overall, cleanly & systematically separating them by stdlib, exceptions, (something else?) sounds like a great move.

But I also wouldn't want to block the MINGW support on making this too big a project.

Reply:
From what is mentioned above, it's rather bad to export something that's in the pybind11:: namespace. Actually, what I would have though from looking around, it's that we are exporting a class that inherits from something that is not exported, which is usually disallowed, but is "safe" if that something is a standard library object.

Regardless, exporting this is "okay" in a very special situations, but MinGW doesn't understand this. So for the exception only, we need to avoid exporting this for now (to restore 2.6 behavior on MinGW). So PYBIND11_EXPORT would remain the same, but PYBIND11_EXPORT_EXCEPTION would only export on non-MinGW.

Reply:
I don't understand all the nuances, but IIUC the upshot is that we'd have `PYBIND11_EXPORT` as before, `PYBIND11_EXPORT_EXCEPTION` in addition, each only defined once in one location, then we replace a few `PYBIND11_EXPORT` with `PYBIND11_EXPORT_EXCEPTION`? If that's true: looks good to me.